hazelcast/src/test/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImplTest.java

```diff
@@ -0,0 +1,487 @@
+        AtomicInteger initialEventCount = new AtomicInteger();
+        ClientClusterServiceImpl clusterService = new ClientClusterServiceImpl(mock(ILogger.class));
+        UUID clusterUuid = UUID.randomUUID();
+        clusterService.handleMembersViewEvent(1, asList(member("127.0.0.1")), clusterUuid);
+        clusterService.addMembershipListener(new InitialMembershipListener() {
+            @Override
+            public void init(InitialMembershipEvent event) {
+                initialEventCount.incrementAndGet();
+            }
+
+            @Override
+            public void memberAdded(MembershipEvent membershipEvent) {
+            }
+
+            @Override
+            public void memberRemoved(MembershipEvent membershipEvent) {
+            }
+        });
+        assertEquals(1, initialEventCount.get());
+    }
+
+    @Test
+    public void testInitialMembershipListener_BeforeInitialListArrives() {
+        AtomicInteger initialEventCount = new AtomicInteger();
+        ClientClusterServiceImpl clusterService = new ClientClusterServiceImpl(mock(ILogger.class));
+        clusterService.addMembershipListener(new InitialMembershipListener() {
+            @Override
+            public void init(InitialMembershipEvent event) {
+                initialEventCount.incrementAndGet();
+            }
+
+            @Override
+            public void memberAdded(MembershipEvent membershipEvent) {
+            }
+
+            @Override
+            public void memberRemoved(MembershipEvent membershipEvent) {
+            }
+        });
+        clusterService.handleMembersViewEvent(1, asList(member("127.0.0.1")), UUID.randomUUID());
+        assertEquals(1, initialEventCount.get());
+    }
+
+    @Test
+    public void testFireOnlyIncrementalEvents_AfterClusterRestart() {
+        AtomicInteger initialEventCount = new AtomicInteger();
+        AtomicInteger addedCount = new AtomicInteger();
+        AtomicInteger removedCount = new AtomicInteger();
+        ClientClusterServiceImpl clusterService = new ClientClusterServiceImpl(mock(ILogger.class));
+        clusterService.handleMembersViewEvent(1, asList(member("127.0.0.1")), UUID.randomUUID());
+
+        clusterService.addMembershipListener(new InitialMembershipListener() {
+            @Override
+            public void init(InitialMembershipEvent event) {
+                initialEventCount.incrementAndGet();
+            }
+
+            @Override
+            public void memberAdded(MembershipEvent membershipEvent) {
+                addedCount.incrementAndGet();
+            }
+
+            @Override
+            public void memberRemoved(MembershipEvent membershipEvent) {
+                removedCount.incrementAndGet();
+            }
+        });
+
+        //called on cluster restart
+        clusterService.onClusterRestart();
+
+        clusterService.handleMembersViewEvent(1, asList(member("127.0.0.2")), UUID.randomUUID());
```

Comment:
Should we also assert for added members address or UUID to check if the correct member is published in member event?

Reply:
Added member checks here https://github.com/hazelcast/hazelcast/pull/20818/commits/7606fc83d49b6bd2f4f7e2b88e401a63e9d46f29